File: drills/ts/solutions/nc65_min_cost_climbing_stairs__from_the_top.ts
export function minCostClimbingStairs(cost: number[]): number {
  // The same recurrence read the other way: instead of "what did it cost to get
  // here", ask "what will it cost to finish from here". Walking backwards, the
  // answer at each step is its own price plus the cheaper of the two ahead, and
  // the start is the better of the first two.
  let oneAhead = 0;
  let twoAhead = 0;
  for (let i = cost.length - 1; i >= 0; i--) {
    [oneAhead, twoAhead] = [cost[i] + Math.min(oneAhead, twoAhead), oneAhead];
  }
  return Math.min(oneAhead, twoAhead);
}

File: drills/go/solutions/nc65_min_cost_climbing_stairs__from_the_top.go
package main

func minCostClimbingStairs(cost []int) int {
	// Recurse from the top down with a memo: the cost to finish from step
	// i is its own cost plus the cheaper of finishing from i+1 or i+2.
	memo := map[int]int{}
	var from func(i int) int
	from = func(i int) int {
		if i >= len(cost) {
			return 0
		}
		if v, ok := memo[i]; ok {
			return v
		}
		memo[i] = cost[i] + min(from(i+1), from(i+2))
		return memo[i]
	}
	return min(from(0), from(1))
}
